packages/fork-choice/src/protoArray/protoArray.ts

```diff
@@ -218,10 +242,220 @@ export class ProtoArray {
+      proposerBoost: this.previousProposerBoost,
+      justifiedEpoch: this.justifiedEpoch,
+      justifiedRoot: this.justifiedRoot,
+      finalizedEpoch: this.finalizedEpoch,
+      finalizedRoot: this.finalizedRoot,
+      currentSlot,
+    });
+  }
+
+  private getNodeIndexFromLVH(latestValidExecHash: RootHex, ancestorOfIndex: number): number | null {
+    let nodeIndex = this.nodes[ancestorOfIndex].parent;
+    while (nodeIndex !== undefined && nodeIndex >= 0) {
+      const node = this.getNodeFromIndex(nodeIndex);
+      if (
+        (node.executionStatus === ExecutionStatus.PreMerge && latestValidExecHash === ZERO_HASH_HEX) ||
+        node.executionPayloadBlockHash === latestValidExecHash
+      ) {
+        break;
+      }
+      nodeIndex = node.parent;
+    }
+    return nodeIndex !== undefined ? nodeIndex : null;
+  }
+
+  private invalidateNodeByIndex(nodeIndex: number): ProtoNode {
+    const invalidNode = this.getNodeFromIndex(nodeIndex);
+
+    // If node to be invalidated is pre-merge or valid,it is a catastrophe,
+    // and indicates consensus failure and a non recoverable damage.
+    //
+    // There is no further processing that can be done.
+    // Just assign error for marking proto-array perma damaged and throw!
+    if (
+      invalidNode.executionStatus === ExecutionStatus.Valid ||
+      invalidNode.executionStatus === ExecutionStatus.PreMerge
+    ) {
+      const lvhCode =
+        invalidNode.executionStatus === ExecutionStatus.Valid
+          ? LVHExecErrorCode.ValidToInvalid
+          : LVHExecErrorCode.PreMergeToInvalid;
+
+      this.lvhError = {
+        lvhCode,
+        blockRoot: invalidNode.blockRoot,
+        execHash: invalidNode.executionPayloadBlockHash ?? ZERO_HASH_HEX,
+      };
+      throw new ProtoArrayError({
+        code: ProtoArrayErrorCode.INVALID_LVH_EXECUTION_RESPONSE,
+        ...this.lvhError,
+      });
+    }
+
+    invalidNode.executionStatus = ExecutionStatus.Invalid;
+    invalidNode.bestChild = undefined;
+    invalidNode.bestDescendant = undefined;
+
+    return invalidNode;
+  }
+
+  private validateNodeByIndex(nodeIndex: number): ProtoNode {
+    const validNode = this.getNodeFromIndex(nodeIndex);
+    if (validNode.executionStatus === ExecutionStatus.Invalid) {
+      this.lvhError = {
+        lvhCode: LVHExecErrorCode.InvalidToValid,
+        blockRoot: validNode.blockRoot,
+        execHash: validNode.executionPayloadBlockHash,
+      };
+      throw new ProtoArrayError({
+        code: ProtoArrayErrorCode.INVALID_LVH_EXECUTION_RESPONSE,
+        ...this.lvhError,
+      });
+    } else if (validNode.executionStatus === ExecutionStatus.Syncing) {
+      validNode.executionStatus = ExecutionStatus.Valid;
+    }
+    return validNode;
+  }
+
   /**
    * Follows the best-descendant links to find the best-block (i.e., head-block).
    */
   findHead(justifiedRoot: RootHex, currentSlot: Slot): RootHex {
+    if (this.lvhError !== undefined) {
```

Comment:
If someday someone changes lvhError to be null instead of undefined
```suggestion
    if (!this.lvhError) {
```